release/src/Main.hs

```diff
@@ -69,49 +71,59 @@ main = do
+          if null missingDeps
+              then
+                  return []
+              else do
+                  maybeMissingDeps <- forM nonDeployJars $ \a -> do
+                      internalDeps <- bazelQueryDeps (T.unpack (getBazelTarget (artTarget a)))
+                      let missingDeps = filter (`Set.notMember` allMavenTargets) internalDeps
+                      if null missingDeps then return Nothing else return (Just (a, missingDeps))
+                  return $ Maybe.catMaybes maybeMissingDeps
 
-      let onlyMissing = filter (not . null . snd) missingDepsForAllArtifacts
       -- now we can report all the missing dependencies per artifact
-      when (not (null onlyMissing)) $ do
-                  $logError "Some internal dependencies are not published to maven central!"
-                  forM_ onlyMissing $ \(artifact, missingDeps) -> do
-                      $logError (getBazelTarget (artTarget artifact))
-                      forM_ missingDeps $ \dep -> $logError ("\t- "# T.pack dep)
-                  liftIO exitFailure
-
-      mvnFiles <- fmap concat $ forM mvnArtifacts $ \a -> do
-          fs <- artifactFiles a
-          pure $ map (a,) fs
-      mapM_ (\(_, (inp, outp)) -> copyToReleaseDir bazelLocations releaseDir inp outp) mvnFiles
-
-      mvnUploadArtifacts <- concatMapM mavenArtifactCoords mvnArtifacts
-      validateMavenArtifacts releaseDir mvnUploadArtifacts
+      when (not (null missingDepsForAllArtifacts)) $ do
+          $logError "Some internal dependencies are not published to Maven Central!"
+          forM_ missingDepsForAllArtifacts $ \(artifact, missingDeps) -> do
+              $logError (getBazelTarget (artTarget artifact))
+              forM_ missingDeps $ \dep -> $logError ("\t- "# T.pack dep)
+          liftIO exitFailure
 
       -- npm packages we want to publish.
       let npmPackages =
-            [ "//language-support/ts/daml-types"
-            , "//language-support/ts/daml-ledger"
-            , "//language-support/ts/daml-react"
-            ]
+              [ "//language-support/ts/daml-types"
+              , "//language-support/ts/daml-ledger"
+              , "//language-support/ts/daml-react"
+              ]
       -- make sure the npm packages can be build.
       $logDebug "Building language-support typescript packages"
       forM_ npmPackages $ \rule -> liftIO $ callCommand $ "bazel build " <> rule
 
-      if | getPerformUpload optsPerformUpload -> do
+      if  | getPerformUpload optsPerformUpload -> do
+              mvnFiles <- fmap concat $ forM mvnArtifacts $ \a -> do
+                  fs <- artifactFiles a
+                  pure $ map (a,) fs
+              mapM_ (\(_, (inp, outp)) -> copyToReleaseDir bazelLocations releaseDir inp outp) mvnFiles
+
+              mvnUploadArtifacts <- concatMapM mavenArtifactCoords mvnArtifacts
+              validateMavenArtifacts releaseDir mvnUploadArtifacts
```

Comment:
This doesn’t look right to me. Now we are only doing validation when we upload. We want to validate on every PR so we discover issues before we try to make a release.
I’m fine with skipping validation when `optsLocallyInstallJars` is set.

Reply:
Moved it back; I think it's a good idea to do it in every case.

If I ever take a look at speeding up uploading, I think we can probably do it in one step, which means Maven itself will do the validation. (I think.)

Reply:
I wouldn’t spend any time on trying to speed up uploading. It’s just not an issue ime (at least since we killed bintray). Other platforms usually take a bit longer (looking at you Windows) so it runs while we’re waiting for Windows.

Reply:
Cool. I wasn't planning on it. Now I shall be extra sure not to plan on it.